README
== Introduction ==

This is a SDK interface for the oVirt Virtualization.

== Usage ==

http://www.ovirt.org/wiki/SDK#IDE_Examples

CHANGES.adoc
= Changes

This document describes the relevant changes between releases of the SDK.

== 4.4.14 / Jul 6 2021

Bug Fixes:

* backup: use newly added "SUCCEEDED/FAILED" backup phases
  https://bugzilla.redhat.com/1849861

* backup_vm.py: fix download backup disk command

== 4.4.13 / Jun 14 2021

* Upgrade to Model 4.4.31

Bug Fixes:

* Preserve /dev/null when running under root user
  https://bugzilla.redhat.com/1956750

== 4.4.12 / Apr 27 2021

* Upgrade to Model 4.4.29

== 4.4.11 / Apr 20 2021

* Upgrade to Model 4.4.28

== 4.4.10 / Feb 7 2021

* Upgrade to Model 4.4.25

== 4.4.9 / Dec 17 2020

* Upgrade to Model 4.4.22

== 4.4.8 / Dec 7 2020

* Upgrade to Model 4.4.21

== 4.4.7 / Nov 2 2020

* Upgrade to Metamodel 1.3.4

Bug Fixes:

* Support enum URL parameters

== 4.4.6 / Oct 20 2020

* Upgrade to Model 4.4.20

== 4.4.5 / Oct 13 2020

* Upgrade to Model 4.4.19

* Upgrade to Metamodel 1.3.3

* Upgrade to Java 11

* Update build scripts and remove old ones

* Use helper classes in example scripts


== 4.4.4 / Jul 1 2020

* Update to model 4.4.17

* Improve examples code around image transfer features

== 4.4.3 / May 18 2020

* Update to model 4.4.16

* Updates to some example scripts


== 4.4.2 / Mar 19 2020

Update to model 4.4.14

New features:

* Make Travis depend on openjdk

== 4.4.1 / Jan 16 2019

New features:

* automation: add EL8

== 4.4.0 / Dec 16 2019
Update to model 4.4.11
move from fc27 to fc30

New features:

* Make build.py script Python 3 compatible

* automation: prepare for EL8 / python3 support

* automation: add standard pipelines conf

Bug Fixes:

* Sort input file list so that ovirtsdk4/xml.so builds in a reproducible way
  http://bugs.python.org/issue30461

* packaging: adhere to fedora naming guidelines

Deprecations:

* packaging: drop fc28 spec (EOL)

* automation: drop fc29 (EOL)

== 4.3.1 / Mar 27 2018
Update to model 4.3.22.

New features:

* Add Hosted Engine disk types to content type
  Bug-Url: https://bugzilla.redhat.com/1663626[1663626].

* Add Hosted Engine disk types to content type
  Bug-Url: https://bugzilla.redhat.com/1600788[1600788].

== 4.3.0 / Jan 16 2018
Update to model 4.3.21.

New features:

* Add `activate` input param to add/install/approve host
  http://bugzilla.redhat.com/1561539[1561539].

* Added support for incremental backup.

* Added block size Storage domain property
  https://bugzilla.redhat.com/1592916[1592916].

* Added V5 storage format
  https://bugzilla.redhat.com/1592916[1592916].

* Add driverSensitiveOptions to managed block storage type.

* Add managed block storage type.

* Setup networks commit on success.

* Add vGPU placement to Host
  https://bugzilla.redhat.com/1641125[1641125].

* Specify cloud-init protocol in vm intialization
  https://bugzilla.redhat.com/1611889[1611889].

* Added SATA to DiskInterface enum.

* Add disks link to Snapshot type.

Bug fixes:

* Change HostNic statistics and labels to Link
  https://bugzilla.redhat.com/1661207[1661207].


== 4.2.9 / Oct 4 2018
Update to model 4.2.37.

New features:

*  Add support for OpenStack Identity API v3
   https://bugzilla.redhat.com/1598391[1598391].

*  Added linkedGlusterVolume to the StorageServerConnection
   type and services
   https://bugzilla.redhat.com/1563257[1563257].

== 4.2.8 / Aug 14 2018
Update to model 4.2.36.

New features:

* Add sync all networks in cluster
  https://bugzilla.redhat.com/1558847[1558847].

* Add note to upgrade host
  https://bugzilla.redhat.com/1603020[1603020].

* Indicate network operation in progress
  https://bugzilla.redhat.com/1477599[1477599].

* Poly dhcpv6 and autoconf iface config
  https://bugzilla.redhat.com/1360839[1360839].

* Add 'Multi Queues Enabled' to VmBase
  https://bugzilla.redhat.com/1574771[1574771].

* Added Reduce action to Disk services
  https://bugzilla.redhat.com/1443963[#1443963].

* Add documentation to DiskService
  https://bugzilla.redhat.com/1585572[1585572].

== 4.2.7 / Jun 6 2018
Update to model 4.2.33.

New features:

* Add search parameter to providers
  https://bugzilla.redhat.com/1572071[#1572071].

== 4.2.6 / Apr 17 2018
Update to model 4.2.32.

New features:

* ImageTransfer - adding 'clientInactivityTimeout' property
  https://bugzilla.redhat.com/1563278[#1563278].

* Add architecture to OperatingSystemInfo
  https://bugzilla.redhat.com/1552026[#1552026].

* Add import template from configuration
  https://bugzilla.redhat.com/1526032[#1526032].

== 4.2.5 / Apr 10 2018
Update to model 4.2.31.

New features:

* Add `cancel` action to transfer image session.

* Add `export` template as OVA to a given path on a host
  https://bugzilla.redhat.com/1526033[#1526033].

* Add service to list user groups
  https://bugzilla.redhat.com/1545716[1545716].

* Expose `syncAllNetworks` command for `HostService`.

* Move `VmPlacementPolicy` attribute from Vm to `VmBase`
  https://bugzilla.redhat.com/1375678[#1375678].

* Document `JobsService` as a searchable entity.

* Add `lease` parameter to preview snapshot
  https://bugzilla.redhat.com/1532577[#1532577].

Bug fixes:

* Add reason message to failed request.

* Deprecate `OsType`
  https://bugzilla.redhat.com/1537679[#1537679].

== 4.2.4 / Jan 24 2018
Bug fixes:

* Change pycrul E_OPERATION_TIMEDOUT to E_OPERATION_TIMEOUTED.

== 4.2.3 / Jan 17 2018
Update to model 4.2.28 and metamodel 1.2.16.

New features:

* Deprecate the `discard_zeroes_data` attribute of the `LogicalUnit`
  type and the `supports_discard_zeroes_data` of the `StorageDomain`
  type, as they have been removed from the kernel.

* Add new `content_type` attribute to the `Disk` type
  https://bugzilla.redhat.com/1511420[#1511420].

* Add support for setting external network providers by name when adding
  or updating a cluster https://bugzilla.redhat.com/1511369[#1511369].

* Add new `unmanaged` attribute to the `OpenStackNetworkProvider` type.

* Move virtual NIC profile mappings insde registration configuration
   https://bugzilla.redhat.com/1522799[#1522799].

Bug fixes:


parser_tab.py

# parser_tab.py
# This file is automatically generated. Do not edit.
_tabversion = '3.2'

_lr_method = 'LALR'

_lr_signature = '[2`\x88\xc7\xdc\x98^\xc4\x8e\xf3\xfc9\x9a\r\xc2'

_lr_action_items = {'GT':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 42, 44, 45, 46, ], [-6, -36, -10, -8, -36, -7, -11, -9, 23, -36, -12, -13, 23, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -22, -26, -23, -16, ]), 'WORD':([0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17, 19, 22, 23, 26, 27, 34, 51, ], [2, -34, -6, 9, -5, -1, -35, 2, -10, -8, 9, -7, -11, -33, -2, -9, 32, -4, 36, 36, 36, 32, -3, ]), 'STRING':([2, 3, 9, 10, 11, 12, 13, 17, 19, 23, 26, 27, 34, ], [-6, 13, -10, -8, 13, -7, -11, -9, 33, 37, 37, 37, 33, ]), 'HEREDOC':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45, 46, 50, ], [-6, -36, -10, -8, -36, -7, -11, -9, -36, -36, -12, -13, -36, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, 48, -22, -28, -26, -23, -16, -27, ]), '=':([19, ], [34, ]), 'NEWLINE':([0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, ], [1, -34, -6, -36, -5, -1, 15, 1, -10, -8, -36, -7, -11, 1, -33, -2, -9, -36, -36, -12, -13, -4, -36, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -36, -22, -28, -26, -23, -16, 1, -29, -30, -27, -3, ]), 'LT':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 42, 44, 45, 46, ], [-6, -36, -10, -8, -36, -7, -11, -9, 27, -36, -12, -13, 27, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -22, -26, -23, -16, ]), 'PIPE':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 42, 44, 45, 46, ], [-6, -36, -10, -8, -36, -7, -11, -9, -36, -36, -12, -13, 41, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -22, -26, -23, -16, ]), 'BANG':([0, 1, 4, 6, 7, 8, 15, 16, 22, 51, ], [5, -34, -5, -1, -35, 5, -33, -2, -4, -3, ]), 'SHELL':([5, 41, ], [14, 50, ]), 'OPTION':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 30, 31, 32, 33, 34, 35, 46, ], [-6, -36, -10, -8, 19, -7, -11, -9, 19, -36, -12, -13, -14, -15, -17, -18, -36, -19, -16, ]), 'MARKER':([24, ], [39, ]), 'LTLT':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 42, 44, 45, 46, ], [-6, -36, -10, -8, -36, -7, -11, -9, 24, -36, -12, -13, 24, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -22, -26, -23, -16, ]), ';':([0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, ], [7, -34, -6, -36, -5, -1, -35, 7, -10, -8, -36, -7, -11, 7, -33, -2, -9, -36, -36, -12, -13, -4, -36, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -36, -22, -28, -26, -23, -16, 7, -29, -30, -27, -3, ]), 'GTGT':([2, 3, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 25, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 42, 44, 45, 46, ], [-6, -36, -10, -8, -36, -7, -11, -9, 26, -36, -12, -13, 26, -21, -20, -14, -15, -17, -18, -36, -19, -31, -32, -25, -24, -22, -26, -23, -16, ]), '$end':([1, 4, 6, 7, 8, 15, 16, 22, 51, ], [-34, -5, -1, -35, 0, -33, -2, -4, -3, ]), }

_lr_action = { }
for _k, _v in _lr_action_items.items():
   for _x, _y in zip(_v[0], _v[1]):
      if not _x in _lr_action:  _lr_action[_x] = { }
      _lr_action[_x][_k] = _y
del _lr_action_items

_lr_goto_items = {'option_value':([19, 34, ], [31, 46, ]), 'name':([0, 8, ], [3, 3, ]), 'redirections':([18, ], [25, ]), 'heredoc':([40, ], [47, ]), 'argument':([3, 11, ], [10, 17, ]), 'eol':([0, 8, 14, 47, ], [4, 4, 22, 51, ]), 'option_list':([11, ], [18, ]), 'command':([0, 8, ], [6, 16, ]), 'argument_list':([3, ], [11, ]), 'file':([23, 26, 27, ], [38, 44, 45, ]), 'pipeline':([25, ], [40, ]), 'main':([0, ], [8, ]), 'redirection':([18, 25, ], [28, 42, ]), 'empty':([3, 11, 18, 19, 25, 34, 40, ], [12, 20, 29, 35, 43, 35, 49, ]), 'option':([11, 18, ], [21, 30, ]), }

_lr_goto = { }
for _k, _v in _lr_goto_items.items():
   for _x, _y in zip(_v[0], _v[1]):
       if not _x in _lr_goto: _lr_goto[_x] = { }
       _lr_goto[_x][_k] = _y
del _lr_goto_items
_lr_productions = [
  ("S' -> main", "S'", 1, None, None, None),
  ('main -> command', 'main', 1, 'p_main', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 101),
  ('main -> main command', 'main', 2, 'p_main', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 102),
  ('command -> name argument_list option_list redirections pipeline heredoc eol', 'command', 7, 'p_command', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 112),
  ('command -> BANG SHELL eol', 'command', 3, 'p_command', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 113),
  ('command -> eol', 'command', 1, 'p_command', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 114),
  ('name -> WORD', 'name', 1, 'p_name', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 129),
  ('argument_list -> empty', 'argument_list', 1, 'p_argument_list', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 133),
  ('argument_list -> argument', 'argument_list', 1, 'p_argument_list', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 134),
  ('argument_list -> argument_list argument', 'argument_list', 2, 'p_argument_list', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 135),
  ('argument -> WORD', 'argument', 1, 'p_argument', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 146),
  ('argument -> STRING', 'argument', 1, 'p_argument', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 147),
  ('option_list -> empty', 'option_list', 1, 'p_option_list', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 152),
  ('option_list -> option', 'option_list', 1, 'p_option_list', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 153),
  ('option_list -> option_list option', 'option_list', 2, 'p_option_list', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 154),
  ('option -> OPTION option_value', 'option', 2, 'p_option', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 166),
  ('option -> OPTION = option_value', 'option', 3, 'p_option', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 167),
  ('option_value -> WORD', 'option_value', 1, 'p_option_value', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 175),
  ('option_value -> STRING', 'option_value', 1, 'p_option_value', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 176),
  ('option_value -> empty', 'option_value', 1, 'p_option_value', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 177),
  ('redirections -> empty', 'redirections', 1, 'p_redirections', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 182),
  ('redirections -> redirection', 'redirections', 1, 'p_redirections', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 183),
  ('redirections -> redirections redirection', 'redirections', 2, 'p_redirections', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 184),
  ('redirection -> LT file', 'redirection', 2, 'p_redirection', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 195),
  ('redirection -> LTLT MARKER', 'redirection', 2, 'p_redirection', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 196),
  ('redirection -> GT file', 'redirection', 2, 'p_redirection', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 197),
  ('redirection -> GTGT file', 'redirection', 2, 'p_redirection', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 198),
  ('pipeline -> PIPE SHELL', 'pipeline', 2, 'p_pipeline', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 203),
  ('pipeline -> empty', 'pipeline', 1, 'p_pipeline', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 204),
  ('heredoc -> HEREDOC', 'heredoc', 1, 'p_heredoc', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 210),
  ('heredoc -> empty', 'heredoc', 1, 'p_heredoc', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 211),
  ('file -> WORD', 'file', 1, 'p_file', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 216),
  ('file -> STRING', 'file', 1, 'p_file', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 217),
  ('eol -> ; NEWLINE', 'eol', 2, 'p_eol', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 222),
  ('eol -> NEWLINE', 'eol', 1, 'p_eol', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 223),
  ('eol -> ;', 'eol', 1, 'p_eol', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 224),
  ('empty -> <empty>', 'empty', 0, 'p_empty', '/home/mpastern/Coding/mpastern@git.engineering/ovirt-engine-cli-poc (geert)/lib/cli/parser.py', 229),
]
